File: cpp/maxAreaOfIsland.cpp
#include <vector>
#include <queue>
#include <tuple>
#include <set>

using namespace std;


class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int m = 0;
        for (int r = 0; r < grid.size(); r++) {
            for (int c = 0; c < grid[0].size(); c++) {
                if (!grid[r][c]) continue;                
                m = max(dfs(make_tuple(r, c), grid), m);
            }
        }
        return m;
    }

    vector<tuple<int, int>> neighbors(int r, int c, vector<vector<int>> grid) {
        vector<tuple<int, int>> p = {make_tuple(r + 1, c), make_tuple(r - 1, c), make_tuple(r, c + 1), make_tuple(r, c - 1)};
        vector<tuple<int, int>> ret;

        for (tuple<int,int> t : p) {
            if (get<0>(t) >= 0 && get<0>(t) < grid.size() && get<1>(t) >= 0 && get<1>(t) < grid[0].size()) {
                ret.push_back(t);
            }
        }
        return ret;   
    }

    int dfs(tuple<int, int> root, vector<vector<int>>& grid) {
        queue<tuple<int, int>> q;
        int total = 0;

        q.push(root);  
        grid[get<0>(root)][get<1>(root)] = 0;

        while (!q.empty()) {
            tuple<int, int> c = q.front();
            q.pop();
            total += 1;

            for (tuple<int, int> t : neighbors(get<0>(c), get<1>(c), grid)) {
                if (!grid[get<0>(t)][get<1>(t)]) continue;
                q.push(t);
                grid[get<0>(t)][get<1>(t)] = 0;
            }
        }
        return total;
    }
};

File: py/maxAreaOfIsland.py
class Solution:
    def maxAreaOfIsland(self, grid: list[list[int]]) -> int:
        m = 0
        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if not grid[r][c]:
                    continue
                m = max(m, Solution.dfs([r, c], grid))
        return m
    
    def neighbors(r, c, grid):
        return [[x,y] for x, y in [[r + 1, c], [r - 1, c], [r, c + 1], [r, c - 1]] if 0 <= x < len(grid) and 0 <= y < len(grid[0])]


    def dfs(coord:list[int], grid):
        total = 0
        q = []
        q.append(coord)
        grid[coord[0]][coord[1]] = 0


        while q:
            cx, cy = q.pop()
            total +=1

            for nx, ny in Solution.neighbors(cx, cy, grid):
                if grid[nx][ny] == 1:
                    q.append([nx, ny])
                    grid[nx][ny] = 0
        
        return total



